src/lexer.c

```diff
@@ -1,109 +1,130 @@
+ * @brief Contains functions to create a double-linked list of tokens.
+ */
+
+#include "mod_lexer.h"
+/* #include "lexer_list.h"
+#include "lexer_utils.h"
+#include "lexer_tok_utils.h" */
 
 /**
- * @brief Iterates through input string and categorises character sequences
- * into tokens.
- * 
- * Spaces are being skipped.
- * The input is divided into <simple commands> which are separated by
- * pipes '|' and may consist of the strings
- * 
- * 		<infile exe opt1 opt2 opt3 >outfile
- * 
- * Each of these input strings represent a token. These tokens are saved
- * according to each simple command. 
- * 
- * @param data 	Overarching struct to hold all necessarry data.
- * @param input Input string delivered by <readline>.
- * @return t_err 
+ * @brief Create a list of token.
+ *
+ * @param lst_head	Head node of token list.
+ * @param input 	Input string to be tokenised.
+ * @return t_err 	ERR_EMPTY, ERR_MALLOC, SUCCESS
  */
-t_err	ft_lexer(t_data *data, char *input)
+t_err	ft_lex_input(t_tkn_list	**lst_head, char *input)
 {
-	int		i;
-	bool	exe_found;
+	t_src	src;
+	t_err	err;
+	t_tok	token;	
 
-	data->token = NULL;
-	i = 0;
-	while (input[i])
+	ft_init_lexer(&src, input);
+	err = ft_tokenise(&src, &token);
+	while (err != ERR_EOF || !*lst_head)
 	{
-		exe_found = false;
-		i += ft_skip_space(input);
-		while (input[i] != '|')
+		err = ft_new_node(lst_head, token.str);
+		if (err != SUCCESS)
 		{
-			data->err = ft_categorise(data, input + i, &exe_found);
-			if (data->err != SUCCESS)
-				return (data->err);
-			while (input[i] != ' ' && input[i] != '|')
-				i++;
-			i += ft_skip_space(input + i);
+			ft_free_lst(lst_head);
+			ft_free_tok(&token);
+			return (err);
 		}
-		i++;
+		err = ft_tokenise(&src, &token);
 	}
 	return (SUCCESS);
 }
 
+/**
+ * @brief Initate the lexer.
+ *
+ * Fills source struct with values from the input string.
+ * @param src 		Struct containing data on input/source string.
+ * @param input 	Input string.
+ */
+void	ft_init_lexer(t_src *src, char *input)
+{
+	src->buf = input;
+	src->buf_size = ft_strlen(input);
+	src->cur_pos = INIT_SRC_POS;
+}
 
-t_err	ft_categorise(t_data *data, char *input, bool *exe_found)
+/**
+ * @brief Create a new node and add it to the end of the token list.
+ *
+ * @param lst_head	Head node of token list.
+ * @param content	Content string to be written into new node.
+ * @return t_err 	ERR_MALLOC, SUCCESS
+ */
+t_err	ft_new_node(t_tkn_list	**lst_head, char *content)
 {
-	int		i;
+	t_tkn_list	*new;
 
-	i = 0;
-	if (input[i] == '<')
-	{
-		if (input[i + 1] == '<')
-			data->err = ft_save_heredoc(data, input);
-		else
-			data->err = ft_save_infile(data, input);
-	}
-	else if (input[i] == '>')
-	{
-		if (input[i + 1] == '>')
-			data->err = ft_save_append(data, input);
-		else
-			data->err = ft_save_outfile(data, input);
-	}
-	else
-		data->err = ft_categorise2(data, input, exe_found);
-	if (data->err != SUCCESS)
-		return (data->err);
+	new = (t_tkn_list *)malloc(sizeof(t_tkn_list));
```

Comment:
why not? that's actually how the official definition is afaik. We just leave it away out of convenience as it is implicitly assumed by the compiler

Reply:
https://stackoverflow.com/questions/605845/do-i-cast-the-result-of-malloc 

I like this one for answers. They basically boil down to it's not necessary and cluttering. 
But there are also some peeps which prefer to cast.
For my money it should be consistent. Since we generally don't cast the return, I think it's confusing to cast it here.

Reply:
Further down in the same page there is this which I find quite appealing :)

As with assertions, annotations that reveal intent decrease bugs. It forces an extra cognitive cross-check. 

But I agree that we should keep it consistent. As it would be more laborous to alter all instances where the cast has been omitted, we will just leave it away here.